util/rateLimit.go

```diff
@@ -0,0 +1,461 @@
+	if cm.running {
+		return
+	}
+	cm.running = true
+	go cm.run(ctx, wg)
+}
+
+func (cm *redCongestionManager) run(ctx context.Context, wg *sync.WaitGroup) {
+	tick := 0
+	targetRate := float64(0)
+	consumedByClient := map[ErlClient]*[]time.Time{}
+	serves := []time.Time{}
+	for {
+		cutoff := time.Now().Add(-1 * cm.window)
+		// first process any new events happening
+		select {
+		case e := <-cm.consumed:
+			if consumedByClient[e.c] == nil {
+				ts := []time.Time{}
+				consumedByClient[e.c] = &ts
+			}
+			*(consumedByClient[e.c]) = append(*(consumedByClient[e.c]), e.t)
+			prune(consumedByClient[e.c], cutoff)
+			tick = (tick + 1) % cm.targetRateRefreshTicks
+		case e := <-cm.served:
+			serves = append(serves, e.t)
+			tick = (tick + 1) % cm.targetRateRefreshTicks
+		default:
+		}
+		// only recalculate the service rate every N ticks, because we have to make sure all entries are pruned
+		if tick == 0 {
+			tick = 1
+			prune(&serves, cutoff)
+			for c := range consumedByClient {
+				if prune(consumedByClient[c], cutoff) == 0 {
+					delete(consumedByClient, c)
+				}
+			}
+			targetRate = 0
+			// targetRate is the average service rate per client per second
+			if len(consumedByClient) > 0 {
+				targetRate = float64(len(serves)) / float64(len(consumedByClient)) / float64(cm.window/time.Second)
+			}
+		}
+		// consume all queries
+		for {
+			select {
+			case query := <-cm.shouldDropQueries:
+				query.ret <- cm.shouldDrop(targetRate, query.c, consumedByClient[query.c])
+			default:
+				goto GOTNONE
+			}
+		}
+	GOTNONE:
+
+		// check for context Done and close the thread. Probably only needed for tests
+		// record the latest data to the struct for tests
+		select {
+		case <-ctx.Done():
+			// prune lists and calculate targetRate once more before writing and exiting
+			prune(&serves, cutoff)
+			for c := range consumedByClient {
+				if prune(consumedByClient[c], cutoff) == 0 {
+					delete(consumedByClient, c)
+				}
+			}
+			targetRate = 0
+			// targetRate is the average service rate per client per second
+			if len(consumedByClient) > 0 {
+				targetRate = float64(len(serves)) / float64(len(consumedByClient)) / float64(cm.window/time.Second)
+			}
+			cm.setTargetRate(targetRate)
+			cm.setConsumedByClient(consumedByClient)
+			cm.setServes(serves)
+			cm.stop(wg)
+			return
+		default:
+		}
+	}
+}
+
+func (cm *redCongestionManager) setTargetRate(tr float64) {
+	cm.targetRate = tr
+}
+
+func (cm *redCongestionManager) setConsumedByClient(cbc map[ErlClient]*[]time.Time) {
+	cm.consumedByClient = cbc
+}
+
+func (cm *redCongestionManager) setServes(ts []time.Time) {
+	cm.serves = ts
+}
+
+func (cm *redCongestionManager) stop(wg *sync.WaitGroup) {
+	cm.runLock.Lock()
+	defer cm.runLock.Unlock()
+	cm.running = false
+	if wg != nil {
+		wg.Done()
+	}
+}
+
+func prune(ts *[]time.Time, cutoff time.Time) int {
+	i := 0
+	// prune times from the given list by pulling in any times which are recent
+	for _, t := range *ts {
```

Comment:
this is the most correct way. 
But you can make it a lot cheapear by finding the first `i` that is after `cutoff` and keeping `ts[i:]`. 

Data is pushed in temporal order, that means once you see a value that is youn enough.
All the remaining items will be. 

Yes, there is a slight chance of a race condition between threads inserting maing 2 items be potentially not follow this, but it doesnt matter, they will be close enough that the extra event can fall under time rounding error of the window.

Reply:
I agree, and will be implementing this, but thinking out loud -- does it cost more memory with this faster way? since we're reliably moving the base index of the list forward, won't that cause new memory to get allocated more regularly?